### PR TITLE
Closest nodes feature

### PR DESCRIPTION
**Pull request by andrewosh**
This PR adds the ability to pass closest nodes into condition functions in `put` and `del`. It also lets you optionally return the closest node from `get`.

This feature is useful mostly for mounting sub-tries within a hypertrie (which is implemented in a separate module).

### CODE REVIEW

File: lib/put.js [hunk 1]
```diff
@@ -2,12 +2,27 @@ const Node = require('./node')
...
 
-function Put (db, key, value, { batch, del, condition = null, hidden = false, valueBuffer = null }, cb) {
+function putDefaultOptions (opts) {
+  return {
+    condition: null,
+    closest: false,
+    hidden: false,
+    valueBuffer: null,
+    flags: 0,
+    ...opts
```
Comment (mafintosh) [inline lib/put.js]: prob breaks our travis for non-spread node support

File: lib/node.js [hunk 2]
```diff
@@ -7,24 +7,36 @@ const KEY = Buffer.alloc(16)
...
+Node.Flags = Flags
 
 Node.prototype[inspect] = function (depth, opts) {
   return ((opts && opts.stylize) || defaultStylize)({seq: this.seq, key: this.key, value: this.value}, 'object')
 }
 
+Object.defineProperty(Node.prototype, 'hidden', {
+  enumerable: true,
+  get: function () {
+    return this.flags && this.flags & Flags.HIDDEN
```
Comment (mafintosh) [inline lib/node.js]: you don't need to guard against .flags. also return !!(this.flags & Flags.HIDDEN)

File: lib/node.js [hunk 3]
```diff
@@ -37,12 +49,12 @@ Node.prototype.path = function (i) {
 Node.prototype.final = function () {
   if (!this.valueBuffer || this.value !== null) return this
   this.value = this.valueEncoding ? this.valueEncoding.decode(this.valueBuffer) : this.valueBuffer
+  this.flags = this.flags >> 8
```
Comment (mafintosh) [inline lib/node.js]: needs a comment

Comment (andrewosh) [reply lib/node.js]: Added a comment explaining that this is required to support user-defined flags while hiding internal flags.

File: lib/put.js (hunk 1 above)
Comment (andrewosh) [reply lib/put.js]: `...` -> `Object.assign`

**Comment by mafintosh** [comment]
3.6.0

**Pull request merged**